File: .tmp_sc4013/java_22/3967_finish_time_of_tasks_ii.java
// CONFIG class=Solution method=minFinishTime types=None
// LeetCode 3967 - Finish Time of Tasks II
// https://leetcode.com/problems/finish-time-of-tasks-ii/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    static class Edge {
        int to, reverse;
        Edge(int to, int reverse) { this.to = to; this.reverse = reverse; }
    }

    private static long combine(long minimum, long maximum, int count, int base) {
        if (count == 0) return base;
        return 2 * maximum - minimum + base;
    }

    public long minFinishTime(int n, int[][] edges, int[] baseTime) {
        List<Edge>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            int iu = graph[u].size(), iv = graph[v].size();
            graph[u].add(new Edge(v, iv));
            graph[v].add(new Edge(u, iu));
        }
        int[] parent = new int[n], parentEdge = new int[n];
        Arrays.fill(parent, -2);
        parent[0] = -1;
        List<Integer> order = new ArrayList<>();
        order.add(0);
        for (int i = 0; i < order.size(); i++) {
            int u = order.get(i);
            for (Edge edge : graph[u]) {
                if (parent[edge.to] == -2) {
                    parent[edge.to] = u;
                    parentEdge[edge.to] = edge.reverse;
                    order.add(edge.to);
                }
            }
        }
        long[][] incoming = new long[n][];
        for (int i = 0; i < n; i++) incoming[i] = new long[graph[i].size()];
        for (int oi = n - 1; oi > 0; oi--) {
            int u = order.get(oi);
            long minimum = 1L << 62, maximum = -1;
            int count = 0;
            for (int edgeIndex = 0; edgeIndex < incoming[u].length; edgeIndex++) {
                if (edgeIndex == parentEdge[u]) continue;
                long value = incoming[u][edgeIndex];
                minimum = Math.min(minimum, value);
                maximum = Math.max(maximum, value);
                count++;
            }
            long value = combine(minimum, maximum, count, baseTime[u]);
            int parentNode = parent[u];
            int reverseIndex = graph[u].get(parentEdge[u]).reverse;
            incoming[parentNode][reverseIndex] = value;
        }
        long answer = 1L << 62;
        for (int u : order) {
            long min1 = 1L << 62, min2 = 1L << 62;
            int minIndex = -1;
            long max1 = -1, max2 = -1;
            int maxIndex = -1;
            for (int i = 0; i < incoming[u].length; i++) {
                long value = incoming[u][i];
                if (value < min1) {
                    min2 = min1;
                    min1 = value;
                    minIndex = i;
                } else if (value < min2) min2 = value;
                if (value > max1) {
                    max2 = max1;
                    max1 = value;
                    maxIndex = i;
                } else if (value > max2) max2 = value;
            }
            long rootValue = combine(min1, max1, graph[u].size(), baseTime[u]);
            answer = Math.min(answer, rootValue);
            for (int i = 0; i < graph[u].size(); i++) {
                Edge edge = graph[u].get(i);
                if (edge.to == parent[u]) continue;
                if (graph[u].size() == 1) {
                    incoming[edge.to][edge.reverse] = baseTime[u];
                    continue;
                }
                long minimum = min1, maximum = max1;
                if (i == minIndex) minimum = min2;
                if (i == maxIndex) maximum = max2;
                incoming[edge.to][edge.reverse] = combine(minimum, maximum, graph[u].size() - 1, baseTime[u]);
            }
        }
        return answer;
    }
}

File: .tmp_rs4013/rust_check22/3967_finish_time_of_tasks_ii.rs
struct Solution;
// LeetCode 3967 - Finish Time of Tasks II
// https://leetcode.com/problems/finish-time-of-tasks-ii/

struct Edge {
    to: usize,
    reverse: usize,
}

fn combine(minimum: i64, maximum: i64, count: i32, base: i32) -> i64 {
    if count == 0 {
        return base as i64;
    }
    2 * maximum - minimum + base as i64
}

impl Solution {
    pub fn min_finish_time(n: i32, edges: Vec<Vec<i32>>, base_time: Vec<i32>) -> i64 {
        let n = n as usize;
        let mut graph: Vec<Vec<Edge>> = (0..n).map(|_| Vec::new()).collect();
        for edge in &edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            let iu = graph[u].len();
            let iv = graph[v].len();
            graph[u].push(Edge { to: v, reverse: iv });
            graph[v].push(Edge { to: u, reverse: iu });
        }
        let mut parent = vec![-2i32; n];
        let mut parent_edge = vec![0usize; n];
        parent[0] = -1;
        let mut order = vec![0usize];
        let mut i = 0;
        while i < order.len() {
            let u = order[i];
            for edge in &graph[u] {
                if parent[edge.to] == -2 {
                    parent[edge.to] = u as i32;
                    parent_edge[edge.to] = edge.reverse;
                    order.push(edge.to);
                }
            }
            i += 1;
        }
        let mut incoming = vec![Vec::new(); n];
        for i in 0..n {
            incoming[i] = vec![0i64; graph[i].len()];
        }
        for oi in (1..n).rev() {
            let u = order[oi];
            let mut minimum = 1i64 << 62;
            let mut maximum = -1i64;
            let mut count = 0;
            for edge_index in 0..incoming[u].len() {
                if edge_index == parent_edge[u] {
                    continue;
                }
                let value = incoming[u][edge_index];
                minimum = minimum.min(value);
                maximum = maximum.max(value);
                count += 1;
            }
            let value = combine(minimum, maximum, count, base_time[u]);
            let parent_node = parent[u] as usize;
            let reverse_index = graph[u][parent_edge[u]].reverse;
            incoming[parent_node][reverse_index] = value;
        }
        let mut answer = 1i64 << 62;
        for &u in &order {
            let mut min1 = 1i64 << 62;
            let mut min2 = 1i64 << 62;
            let mut min_index = -1i32;
            let mut max1 = -1i64;
            let mut max2 = -1i64;
            let mut max_index = -1i32;
            for i in 0..incoming[u].len() {
                let value = incoming[u][i];
                if value < min1 {
                    min2 = min1;
                    min1 = value;
                    min_index = i as i32;
                } else if value < min2 {
                    min2 = value;
                }
                if value > max1 {
                    max2 = max1;
                    max1 = value;
                    max_index = i as i32;
                } else if value > max2 {
                    max2 = value;
                }
            }
            let root_value = combine(min1, max1, graph[u].len() as i32, base_time[u]);
            answer = answer.min(root_value);
            for i in 0..graph[u].len() {
                let edge_to = graph[u][i].to;
                let edge_reverse = graph[u][i].reverse;
                if edge_to as i32 == parent[u] {
                    continue;
                }
                if graph[u].len() == 1 {
                    incoming[edge_to][edge_reverse] = base_time[u] as i64;
                    continue;
                }
                let mut minimum = min1;
                let mut maximum = max1;
                if i as i32 == min_index {
                    minimum = min2;
                }
                if i as i32 == max_index {
                    maximum = max2;
                }
                incoming[edge_to][edge_reverse] =
                    combine(minimum, maximum, graph[u].len() as i32 - 1, base_time[u]);
            }
        }
        answer
    }
}

fn main() {}
